File: javascript/lesson 10/fibonacciNumberRecursive.js
// https://leetcode.com/problems/fibonacci-number/

function fib(n) {
    if (n <= 1) return n;
    return fibo(n, [0, 1]);
}

function fibo(n, dp) {
    if (dp[n] !== undefined) return dp[n];
    dp[n] = fibo(n - 1, dp) + fibo(n - 2, dp);
    return dp[n];
}

File: ruby/lesson 10/fibonacci_number_recursive.rb
# https://leetcode.com/problems/fibonacci-number/

def fib(n)
    return n if n <= 1
    fib_impl(n, [0, 1])
end

def fib_impl(n, dp)
    return dp[n] unless dp[n].nil?
    dp[n] = fib_impl(n - 1, dp) + fib_impl(n - 2, dp)
end
